torch/amp/grad_scaling.py

```diff
@@ -0,0 +1,382 @@
+            If supplied, ``scale_scheduler`` should take 2 arguments.  Both will be one-element
+            ``torch.cuda.FloatTensors``.  The first will contain the current scale value.  The second will contain > 0
+            if gradients currently contain inf or nan, and ``0.0`` otherwise.  The scheduler should use these arguments
+            to determine the next recommended scale, and return a one-element ``torch.cuda.FloatTensor`` containing the
+            next recommended scale.
+        skip_if_inf (bool, optional, default=True):  If False, unscale gradients but call ``step()`` blindly without
+            checking if the gradients contain inf or nan.  If you're manually using the same gradient scale every
+            iteration ("static gradient scaling", as opposed to dynamic gradient scaling) and you're fairly
+            certain no inf/nan gradients will be encountered, setting ``skip_if_inf=False`` can provide a minor
+            performance improvement, because for typical optimizers it avoids copying ``found_inf``'s value back to
+            the CPU.
+            Another case where it makes sense to supply ``skip_if_inf=False`` is batch replay, where you replay this
+            iteration until inf/nan-free gradients are created.  By the time you reach the
+            :func:`optimizer.unscale_and_step()` call, you're sure the gradients do not contain inf/nan, so it's safe
+            to elide the check.  For other (typical) use, the default ``skip_if_inf=True`` is strongly recommended.
+        found_inf (torch.cuda.FloatTensor, optional):  If you've already created a 1-element Tensor that contains > 0.0
+            if the optimizer's current gradients have an inf/nan and 0.0 if they don't (by, for example, an earlier
+            call to ``optimizer.check_inf()``) you may pass this as ``found_inf``.  By default, you don't need to pass
+            ``found_inf`` manually.  :func:`unscale_and_step` will inspect the gradients for you, create a
+            ``found_inf`` tensor, and return it.  Leaving ``found_inf=None`` and letting :func:`unscale_and_step`
+            create it for you is typical.
+        scaling_enabled (bool, optional):  If ``False``, :func:`unscale_and_step` simply calls ``self.step()`` and
+            returns ``opt_ret, None, None``, where opt_ret is the default return value of ``self.step()``.
+
+    Returns:
+        A tuple ``opt_ret, found_inf, recommended_scale``.  ``found_inf`` is a ``torch.cuda.FloatTensor`` that
+        contains > 0 if an inf/nan was found during unscaling, and 0.0 otherwise.  ``recommended_scale`` is a
+        ``torch.cuda.FloatTensor`` containing the recommended gradient scale to use next iteration.  ``opt_ret``
+        is the default return value of the underlying ``self.step()`` call.  If ``scaling_enabled=False``, the return
+        values will be ``opt_ret, None, None``.
+    """
+    if not scaling_enabled:
```

Comment:
So I can use closures here but not in the other case? Do we need to support closures at all?

Reply:
The intention was that the `unscale_and_step` and `step_after_unscale` calls would behave identically to the default `optimizer.step` calls when `scaling_enabled=False` (including supporting closures).  But I can error out in all cases if you think that's better.